text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/translation/MiniMessageTranslator.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * <p>An example of how you might construct a translatable component would be:</p>
+ * <pre>{@code
+ * Component.translatable(
+ *   "my.translation.key", // the translation key, you'd return the MiniMessage string by implementing getMiniMessageString
+ *   Component.text("hello"), // available as <arg:0> or <argument:0>
+ *   Argument.string("today", "monday"), // available as <arg:1>, <argument:1> or <today>
+ *   Argument.tag("danger", Tag.styling(NamedTextColor.RED)), // available as <arg:1>, <argument:1> or <red>, can be closed if needed
+ *   Argument.tagResolver(StandardTags.pride()) // you can even add arbitrary tag resolvers!
+ * );
+ * }</pre>
+ *
+ * @see Argument
+ * @see Tag
+ * @see TagResolver
+ * @see Translator
+ * @see GlobalTranslator
+ * @since 4.20.0
+ */
+public abstract class MiniMessageTranslator implements Translator {
+  private final MiniMessage miniMessage;
+
+  /**
+   * Constructor for a MiniMessageTranslator using the default MiniMessage instance.
+   *
+   * @see MiniMessage#miniMessage()
+   * @since 4.20.0
+   */
+  public MiniMessageTranslator() {
+    this(MiniMessage.miniMessage());
+  }
+
+  /**
+   * Constructor for a MiniMessageTranslator using a specific MiniMessage instance.
+   *
+   * @param miniMessage the MiniMessage instance
+   * @see MiniMessage#miniMessage()
+   * @since 4.20.0
+   */
+  public MiniMessageTranslator(final @NotNull MiniMessage miniMessage) {
+    this.miniMessage = Objects.requireNonNull(miniMessage, "miniMessage");
+  }
+
+  /**
+   * Returns a raw MiniMessage string for the given key.
+   *
+   * <p>If no string is found for the given key, returning {@code null} will use the
+   * {@link TranslatableComponent#fallback() translatable component's fallback} (or the
+   * key itself).</p>
+   *
+   * @param key the key
+   * @param locale the locale
+   * @return the resulting MiniMessage string
+   * @since 4.20.0
+   */
+  @SuppressWarnings("checkstyle:MethodName")
+  protected abstract @Nullable String getMiniMessageString(final @NotNull String key, final @NotNull Locale locale);
+
+  @Override
+  public final @Nullable MessageFormat translate(final @NotNull String key, final @NotNull Locale locale) {
+    return null;
+  }
+
+  @Override
+  public final @Nullable Component translate(final @NotNull TranslatableComponent component, final @NotNull Locale locale) {
+    final String miniMessageString = this.getMiniMessageString(component.key(), locale);
+
+    if (miniMessageString == null) {
+      return null;
+    }
+
+    final Component resultingComponent;
+
+    if (component.arguments().isEmpty()) {
+      resultingComponent = this.miniMessage.deserialize(miniMessageString);
+    } else {
+      resultingComponent = this.miniMessage.deserialize(miniMessageString, new ArgumentTag(component.arguments()));
+    }
+
+    if (component.children().isEmpty()) {
+      return resultingComponent;
+    } else {
+      return resultingComponent.children(component.children());
+    }
```

Comment:
This should be using the new multi-component append method added in #1181 in order to fix #974. Currently, this wipes the children of the resulting component.